backend/src/replication/StatisticsReplication.ts

```diff
@@ -0,0 +1,217 @@
+
+      const intervals = [              // [start,               end,                 label ]
+                                          [now - day,           now - 60,            '24h']       , // from 24 hours ago to now = 1 minute granularity
+        startTime < now - day ?           [now - day * 7,       now - day,           '1w' ] : null, // from 1 week ago to 24 hours ago = 5 minutes granularity
+        startTime < now - day * 7 ?       [now - day * 30,      now - day * 7,       '1m' ] : null, // from 1 month ago to 1 week ago = 30 minutes granularity
+        startTime < now - day * 30 ?      [now - day * 90,      now - day * 30,      '3m' ] : null, // from 3 months ago to 1 month ago = 2 hours granularity
+        startTime < now - day * 90 ?      [now - day * 180,     now - day * 90,      '6m' ] : null, // from 6 months ago to 3 months ago = 3 hours granularity
+        startTime < now - day * 180 ?     [now - day * 365 * 2, now - day * 180,     '2y' ] : null, // from 2 years ago to 6 months ago = 8 hours granularity
+        startTime < now - day * 365 * 2 ? [startTime,           now - day * 365 * 2, 'all'] : null, // from start of statistics to 2 years ago = 12 hours granularity   
+      ];
+
+      for (const interval of intervals) {
+        if (!interval) {
+          continue;
+        }
+        missingStatistics[interval[2] as string] = await this.$getMissingStatisticsInterval(interval, startTime);
+      }
+      
+      return missingStatistics;
+    } catch (e: any) {
+      logger.err(`Cannot fetch missing statistics times from db. Reason: ` + (e instanceof Error ? e.message : e));
+      throw e;
+    }
+  }
+
+  private async $getMissingStatisticsInterval(interval: any, startTime: number): Promise<number[]> {
+    try {
+      const start = interval[0];
+      const end = interval[1];
+      const step = steps[interval[2]];
+
+      const [rows]: any[] = await DB.query(`
+        SELECT UNIX_TIMESTAMP(added) as added
+        FROM statistics
+        WHERE added >= FROM_UNIXTIME(?) AND added <= FROM_UNIXTIME(?)
+      `, [start, end]);
+      const timesAlreadyHere = rows.map(row => row.added);
+
+      const missingTimes: number[] = [];
+
+      if (timesAlreadyHere.length === 0) {
+        for (let time = Math.max(startTime, start); time < end; time += step) {
+            missingTimes.push(time);
+        }
+        return missingTimes;
+      }
+
+      for (let time = Math.max(startTime, start); time < end; time += step) {
+        const closest = this.getClosestElement(time, timesAlreadyHere);
+        if (Math.abs(time - closest.value) > step && !this.skip.has(time)) { 
+          missingTimes.push(time);
+        }
+      }
+
+      return missingTimes;
+    } catch (e: any) {
+      logger.err(`Cannot fetch missing statistics times from db. Reason: ` + (e instanceof Error ? e.message : e));
+      throw e;
+    }
+  }
```

Comment:
It feels like there should be a cleaner/more efficient way to find the missing points than iterating over the entire list of data each time with `getClosestElement(...)`.

My instinct would be to round everything off to the nearest `step`, throw it in a Set or a dictionary, and then do a single pass over the expected *rounded* values to see what's missing.

For the longer intervals, you can probably get the points out of the DB pre-rounded to the correct `step` intervals by using the same `GROUP BY` logic as the statistics APIs you'll be hitting later to fetch the missing data:

https://github.com/mempool/mempool/blob/12d212b89f4a57948bf34ae01943d30ce09ecafa/backend/src/api/statistics/statistics-api.ts#L269-L271